spark3/src/main/java/org/apache/iceberg/spark/procedures/AddFilesProcedure.java

```diff
@@ -0,0 +1,319 @@
+    CatalogTable catalogTable = SparkTableUtil.getCatalogTable(spark(), sourceTable);
+    List<SparkTableUtil.SparkPartition> partitions;
+
+    if (catalogTable.partitionColumnNames().isEmpty()) {
+      // Unpartitioned table
+      if (!partition.isEmpty()) {
+        throw new IllegalArgumentException("Cannot import files from a unpartitioned Hive table when a partition " +
+            "is specified");
+      }
+      Option<String> format =
+          catalogTable.storage().serde().nonEmpty() ? catalogTable.storage().serde() : catalogTable.provider();
+      partitions = ImmutableList.of(new SparkTableUtil.SparkPartition(
+          Collections.emptyMap(),
+          catalogTable.location().toString(),
+          catalogTable.storage().serde().get()));
+
+    } else {
+      // Read partitions Spark via Catalog Interface
+      partitions = filterPartitions(SparkTableUtil.getPartitions(spark(), sourceTable), partition);
+    }
+
+    importPartitions(table, partitions);
+  }
+
+  private void importPartitions(Table table, List<SparkTableUtil.SparkPartition> partitions) {
+    String stagingLocation = table.properties()
+        .getOrDefault(TableProperties.WRITE_METADATA_LOCATION, table.location() + "/metadata");
+    SparkTableUtil.importSparkPartitions(spark(), partitions, table, table.spec(), stagingLocation);
+  }
+
+  private List<SparkTableUtil.SparkPartition> filterPartitions(List<SparkTableUtil.SparkPartition> partitions,
+                                                               Map<String, String> partition) {
+    if (partition.isEmpty()) {
+      // No partition filter arg
+
+      if (partitions.isEmpty()) {
+        throw new IllegalArgumentException("Cannot add files, no files found in the table.");
+      }
+      return partitions;
+    } else {
+      // Partition filter arg passed
+
+      List<SparkTableUtil.SparkPartition> filteredPartitions = partitions
+          .stream().filter(p -> p.getValues().equals(partition)) // Todo Support Wildcards
+          .collect(Collectors.toList());
+
+      if (filteredPartitions.isEmpty()) {
+        throw new IllegalArgumentException(
+            String.format("No partitions found in table for add_file command. " +
+                    "Looking for a partition with the value %s",
+                partition.entrySet().stream().map(e -> e.getKey() + "=" + e.getValue())
+                    .collect(Collectors.joining(",")))
+        );
+      }
+
+      return filteredPartitions;
+    }
+  }
+
+  @Override
+  public String description() {
+    return null;
+  }
+
+  private void validatePartitionSpec(Table table, Map<String, String> partition) {
+    List<PartitionField> partitionFields = table.spec().fields();
+
+    boolean tablePartitioned = !partitionFields.isEmpty();
+    boolean partitionSpecPassed = !partition.isEmpty();
+
+    if (tablePartitioned && partitionSpecPassed) {
```

Comment:
Since we use the map as a filter, we no longer should require the map to contain all arguments. 

This should be valid, I guess:

```
// table partitioned by p1, p2
add_files(
  table => 'db.iceberg_table', -- partitioned Iceberg table
  source_table => '`parquet`.`path/to/table/`'
  partition_filter => map('p1', ‘x’)
)
```

That is instead of using *, which is more flexible.